Adding FunctionCategoryFactory to handle the custom PROCEDURE toolbox category.

AnmAtAnm opened the pull request:
Also, refactored how custom categories are added, making it consistent with how activities (and the activity helper) add mutators and extensions.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/590)
<!-- Reviewable:end -->

RoboErikG commented:
minor nits, then <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java, line 108 at r1](https://reviewable.io:443/reviews/google/blockly-android/590#-KjoRKO2iyUDf4VvgCEi:-KjoRKO2iyUDf4VvgCEj:b-k3m9wa) ([raw file](https://github.com/google/blockly-android/blob/c456e502f7012ff4ba365e706ded3351511b7549/blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java#L108)):*
> ```Java
>     }
> 
>     public static Map<String, CategoryFactory> getToolboxCustomCategories(BlocklyController controller) {
> ```

nit line length

---

*[blocklylib-core/src/main/java/com/google/blockly/model/FunctionCategoryFactory.java, line 18 at r1](https://reviewable.io:443/reviews/google/blockly-android/590#-KjoRVFjZqNZ_Z5M25RC:-KjoRVFkL5B-pZDe7ciX:bikors5) ([raw file](https://github.com/google/blockly-android/blob/c456e502f7012ff4ba365e706ded3351511b7549/blocklylib-core/src/main/java/com/google/blockly/model/FunctionCategoryFactory.java#L18)):*
> ```Java
>     public BlocklyCategory obtainCategory(String customType) {
>         BlockFactory factory = mController.getBlockFactory();
> ```

Should this have a TODO: Implement?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/590#-:-KjoRaYCZNMyc71yzIht:b-7rk1nx)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java, line 108 at r1](https://reviewable.io:443/reviews/google/blockly-android/590#-KjoRKO2iyUDf4VvgCEi:-KjodbyFU6oXR6h5Kc_f:bt2vr1b) ([raw file](https://github.com/google/blockly-android/blob/c456e502f7012ff4ba365e706ded3351511b7549/blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java#L108)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit line length
</blockquote></details>

Fixed.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/FunctionCategoryFactory.java, line 18 at r1](https://reviewable.io:443/reviews/google/blockly-android/590#-KjoRVFjZqNZ_Z5M25RC:-KjodfQAPvmjXF5Z4w_X:bqmjwcz) ([raw file](https://github.com/google/blockly-android/blob/c456e502f7012ff4ba365e706ded3351511b7549/blocklylib-core/src/main/java/com/google/blockly/model/FunctionCategoryFactory.java#L18)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Should this have a TODO: Implement?
</blockquote></details>

Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/590)*
<!-- Sent from Reviewable.io -->
